sridha2412/assignment3/wordsearch.py

```diff
@@ -0,0 +1,124 @@
+class TrieNode:
+    def __init__(self):
+        """ Constructor to initlalize a trie node """
+        self.children = {}
+        self.is_word = False
+
+class Dictionary:
+    def __init__(self, words: list = []):
+        """ Constructor to initlalize a dictionary """
+        self.root = TrieNode()
+        for word in words:
+            self.add_word(word)
+
+    def add_word(self, word: str):
+        """ Method to add a word to dictionary if not already present """
+        current = self.root
+        for char in word:
+            if char not in current.children:
+                current.children[char] = TrieNode()
+            current = current.children[char]
+        current.is_word = True
+
+    def is_prefix(self, prefix: str) -> bool:
+        """ Method to return true if prefix found in dictionary """
+        current = self.root
+        for char in prefix:
+            if char not in current.children:
+                return False
+            current = current.children[char]
+        return True
+
+    def is_word(self, word: str) -> bool:
+        """ Method to return true if word found in dictionary """
+        current = self.root
+        for char in word:
+            if char not in current.children:
+                return False
+            current = current.children[char]
+        return current.is_word
+
+
+class Grid:
+    def __init__(self, grid: list):
+        """ Constructor to initlalize a grid """
+        self.grid = grid
+
+    def get_neighbours(self, x: int, y: int) -> list:
+        """ Method to find all possible neighbouring cells given position (x, y) in a grid """
+        # 8 possible positions - vertical, horizontal and diagonal
+        positions = [(x - 1, y - 1), (x - 1, y), (x - 1, y + 1), (x, y + 1), (x + 1, y + 1), (x + 1, y), (x + 1, y - 1), (x, y - 1)]
+        return [pos for pos in positions if 0 <= pos[0] < len(self.grid) and 0 <= pos[1] < len(self.grid[0])]
+
+    def get_word(self, positions: list) -> str:
+        """ Method to find corresponding word given a list of positions in grid """
+        return ''.join(self.grid[pos[0]][pos[1]] for pos in positions)
+
+    def find_words(self, dictionary: Dictionary, position: list) -> list:
+        """ Method to return all words from given position """
+        words = []
+
+        word = self.get_word(position)
+        if (not dictionary.is_prefix(word)):
+            return []
+        if (dictionary.is_word(word)):
+            words.append(word)
+
+        neighbours = self.get_neighbours(position[-1][0], position[-1][1])
```

Comment:
I do not understand how position[-1][0] can be correct. I thought position was a list of 2 integers?

I now understand reading the usage below: why a list of one tuple where you could simply use the single 2-tuple?

Reply:
it is a list of tuples  so it looks something like [(0, 0), (0, 1), ...]
at each iteration, it appends a position to the tuple to avoid using the same grid 
It is position[-1][0] because I am accessing the last tuple of the list.